server.go

```diff
@@ -352,17 +350,26 @@ func (s *baseServer) handlePacketImpl(p *receivedPacket) bool /* was the packet
 	if !hdr.IsLongHeader {
 		return false
 	}
+	if hdr.Type == protocol.PacketTypeInitial && len(p.data) < protocol.MinInitialPacketSize {
+		s.logger.Debugf("Dropping a packet that is too small to be a valid Initial (%d bytes)", len(p.data))
+		return false
+	}
 	// send a Version Negotiation Packet if the client is speaking a different protocol version
 	if !protocol.IsSupportedVersion(s.config.Versions, hdr.Version) {
 		s.sendVersionNegotiationPacket(p, hdr)
 		return false
 	}
-	if hdr.IsLongHeader && hdr.Type != protocol.PacketTypeInitial {
-		// Drop long header packets.
-		// There's litte point in sending a Stateless Reset, since the client
-		// might not have received the token yet.
-		s.logger.Debugf("Dropping long header packet of type %s (%d bytes)", hdr.Type, len(p.data))
-		return false
+	if hdr.IsLongHeader {
+		if hdr.Type == protocol.PacketType0RTT {
+			s.zeroRTTQueue.Enqueue(hdr.DestConnectionID, p)
```

Comment:
This means you always queue initial packets, even if we already have a session for them?

Reply:
No, if we already have a session, the `packetHandlerMap` will already direct the packets there.

Reply:
Not if you have two go routines at this point in the code in parallel. One creates the session, the other one then ends up queuing a packet which will never be read.

I find the overall structure here somewhat hard to reason about, but I think you need a mutex around checking whether we have a session and queueing, or an additional (possibly noop) dequeue after each queue.